### PR TITLE
Add Windows support

### PR DESCRIPTION
**Pull request by tiangolo**
# Windows support! :tada: 

This uses [ffi](https://github.com/node-ffi/node-ffi/), and some of the familiar packages (`ref` and `ref-wchar`) and adds them as dependencies.

It uses the Windows C++ APIs (via `ffi` and family).

It is all implemented in an additional `windows.js` file, that is required conditionally.

The code is commented in each part to explain how it works, as working with `ffi` and interacting with C++ via Buffers can get a bit tangled (apart from the not-so-good Windows APIs documentation).

---

**Note**: It is based on the code of the PR: https://github.com/sindresorhus/active-win/pull/6 . So it assumes that PR being merged. Nevertheless, it only changes a couple lines from the base code from that PR. So, if we need to change that PR further it won't complicate things too much.

### CODE REVIEW

**Comment by sindresorhus** [comment]
What do you think about using [`fastcall`](https://github.com/cmake-js/fastcall) instead of `node-ffi`? It is faster and seems better maintained.

File: readme.md [hunk 1]
```diff
@@ -45,11 +43,18 @@ Returns the result `Object`.
...
 - `title` - Window title
-- `id` - Window ID
+- `id` - Window ID (in MacOS and Linux)
 - `app` - App owning the window
 - `pid` - Process ID of the app owning the window
 
 
+## Operating system support
+
+It works with MacOS, Linux and Windows (tested in Windows 7 and above).
```
Comment (sindresorhus) [inline readme.md]: MacOS => macOS

---

> Windows (tested in Windows 7 and above)

Just say `Windows 7+`. I'm not interested in supporting anything lower anyways.

File: package.json [hunk 2]
```diff
@@ -39,7 +39,10 @@
     "capture"
   ],
   "dependencies": {
-    "pify": "^2.3.0"
+    "ffi": "^2.2.0",
+    "pify": "^2.3.0",
+    "ref": "^1.3.4",
+    "ref-wchar": "^1.0.2"
```
Comment (sindresorhus) [inline package.json]: Can you make the new dependencies `optionalDependencies`, so it doesn't fail to install on macOS/Linux if something is missing on the user's system to compile.

**Comment by tiangolo** [comment]
I never heard of [fastcall](https://github.com/cmake-js/fastcall) before. Let me try it locally to check that it still works and then all push the update.

File: readme.md (hunk 1 above)
Comment (tiangolo) [inline readme.md]: Cool. Done (will push in a minute).

File: package.json (hunk 2 above)
Comment (tiangolo) [inline package.json]: Done :white_check_mark:

**Comment by sindresorhus** [comment]
Seems like there *is* a way to get the window ID: http://stackoverflow.com/a/3248481/64949

**Comment by tiangolo** [comment]
Hmm, ok. Let me check it, the [documentation for that](https://msdn.microsoft.com/en-us/library/windows/desktop/ms633584(v=vs.85).aspx) doesn't seem to imply quite clearly a window ID. But let me check to see what I can get.

**Comment by sindresorhus** [comment]
> GWL_ID -12 Retrieves the **identifier of the window**.

**Comment by tiangolo** [comment]
Yep, that's what I saw, but following [the example](https://msdn.microsoft.com/en-us/library/windows/desktop/ms632598(v=vs.85).aspx#creating_enumerating_etc) seemed like that could also be an identifier only to the owner or something similar. I don't really know.

Let me code it and see what we get.

---

I'm first trying to use `fastcall`, but it seems it requires more / different dependencies than what I had, so I'm still just trying to "install" it successfully. (Sigh... Windoze :weary:  ).

I'll then try to replace everything I did with `ffi` (it seems it should be "straightforward").

**Comment by tiangolo** [comment]
Well, `fastcall` worked!

---

The `GetWindowLong` is returning me `0` in all the local tests I tried :disappointed: .

---

Thanks for merging https://github.com/sindresorhus/active-win/pull/6 !

I'm not sure what would be the right procedure to update this branch to be based on your latest changes. 

Should I try to rebase or merge?

**Comment by tiangolo** [comment]
As you can just do a squash merge and the PR was already "public", I took the easy / short way and just merged the current `master` into `feature/windows`.

Now I'll check what's happening with the building of `fastcall` in Travis (I got an error already in my version).

I'll create a branch in my fork to fiddle with it and try to solve it and then I'll merge it into this branch.

---

I would like to ask, what do you think we should do with the `GetWindowLong` returning `0`?

The [example](https://msdn.microsoft.com/en-us/library/windows/desktop/ms632598(v=vs.85).aspx#creating_enumerating_etc) makes me think that what the "window ID" refers to is not a global "window ID" but rather a local "window ID", with respect to a "parent" window.

It is used in conjunction with [`EnumChildWindows`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms633494(v=vs.85).aspx):

> Enumerates the child windows that belong to the specified parent window...

And with [`EnumChildProc`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms633493(v=vs.85).aspx):

> An application-defined callback function used with the EnumChildWindows function. ...

Here's the section of the code from the example using exactly the same that we are trying (including the flag) `idChild = GetWindowLong(hwndChild, GWL_ID);`:

```
    LPRECT rcParent; 
    int i, idChild; 
 
    // Retrieve the child-window identifier. Use it to set the 
    // position of the child window. 
 
    idChild = GetWindowLong(hwndChild, GWL_ID); 
 
    if (idChild == ID_FIRSTCHILD) 
        i = 0; 
    else if (idChild == ID_SECONDCHILD) 
        i = 1; 
    else 
        i = 2; 
 
    // Size and position the child window.  
```

It kind of makes me think that it is not what we are expecting / trying to achieve. But still is the closest I've seen to a "window ID".

And well, in my local tests it is currently just returning `0` all the time...

Should we ditch it? What should we do?

**Comment by sindresorhus** [comment]
> Now I'll check what's happening with the building of fastcall in Travis (I got an error already in my version).

Why are you trying to build it on Travis? Shouldn't we use AppVeyor to build it for Windows? Or is it easier to cross-compile it for Windows on Linux?

> Should we ditch it? What should we do?

First we need to think about what the use-case of the window ID is. I would say it's to globally be able to differentiate windows. Maybe we could just create a number based on the process ID and the window ID, to get a globally unique number for a window?

File: windows.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/ms684320(v=vs.85).aspx
+	.function({OpenProcess: ['pointer', ['uint32', 'int', 'uint32']]})
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/ms724211(v=vs.85).aspx
+	.function({CloseHandle: ['int', ['pointer']]})
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/ms684919(v=vs.85).aspx
+	.function({QueryFullProcessImageNameW: ['int', ['pointer', 'uint32', 'pointer', 'pointer']]}));
+
+module.exports = () => {
+	// Windows C++ APIs' functions are declared with capitals, so this rule has to be turned off.
+	/* eslint new-cap: off */
```
Comment (sindresorhus) [inline windows.js]: Use `/* eslint-disable new-cap */`

File: windows.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	const windowTextLength = user32.interface.GetWindowTextLengthW(activeWindowHandle);
+	// Allocate a buffer large enough to hold the window text as "Unicode" (UTF-16) characters (using ref-wchar)
+	// This assumes using the "Basic Multilingual Plane" of Unicode, only 2 characters per Unicode code point
+	// Include some extra bytes for possible null characters
+	const windowTextBuffer = Buffer.alloc((windowTextLength * 2) + 4);
+	// Write the window text to the buffer (it returns the text size, but is not used here)
+	user32.interface.GetWindowTextW(activeWindowHandle, windowTextBuffer, windowTextLength + 2);
+	// Remove trailing null characters
+	const windowTextBufferClean = ref.reinterpretUntilZeros(windowTextBuffer, wchar.size);
+	// The the text as a Javascript string
```
Comment (sindresorhus) [inline windows.js]: `The the text as a Javascript string` => `The text as a JavaScript string`

File: windows.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+const ref = fastcall.ref;
+const wchar = require('ref-wchar');
+
+// Required by QueryFullProcessImageName
+// https://msdn.microsoft.com/en-us/library/windows/desktop/ms684880(v=vs.85).aspx
+const PROCESS_QUERY_LIMITED_INFORMATION = 0x1000;
+
+// Used by GetWindowLong
+// https://msdn.microsoft.com/en-us/library/windows/desktop/ms633584(v=vs.85).aspx
+const GWL_ID = -12;
```
Comment (sindresorhus) [inline windows.js]: Shouldn't we get this global constant instead of hard-coding it?

File: readme.md [hunk 6]
```diff
@@ -45,11 +43,18 @@ Returns the result `Object`.
 ## Result
 
 - `title` - Window title
-- `id` - Window ID
+- `id` - Window ID (in MacOS and Linux)
```
Comment (sindresorhus) [inline readme.md]: `macOS`

File: windows.js (hunk 5 above)
Comment (tiangolo) [inline windows.js]: I understand that we have to hard-code them: https://github.com/node-ffi/node-ffi/issues/369

I know that issue is from `ffi`, not from `fastcall`, but there doesn't seem to be any reference or documentation about it in `fastcall` (or its issues).

File: windows.js (hunk 3 above)
Comment (tiangolo) [inline windows.js]: OK, done.

File: windows.js (hunk 4 above)
Comment (tiangolo) [inline windows.js]: Oops! Thanks.

File: readme.md (hunk 6 above)
Comment (tiangolo) [inline readme.md]: OK :ok_hand:

**Comment by tiangolo** [comment]
> Why are you trying to build it on Travis? Shouldn't we use AppVeyor to build it for Windows? Or is it easier to cross-compile it for Windows on Linux?

I saw some errors in the Travis log because it was not being able to install / build `fastcall`, so I was thinking on trying to debug that just to let Travis finish the build.

But yes, if we were going to test / build for Windows it should be done in AppVeyor.

> First we need to think about what the use-case of the window ID is. I would say it's to globally be able to differentiate windows. Maybe we could just create a number based on the process ID and the window ID, to get a globally unique number for a window?

Hmm, I know that the function we are using returns the thread that created the Window. Maybe we could create an ID based on the PID and thread ID. But if that thread creates more than one Window it wouldn't work as expectect, I think. 

The Window ID that we are getting (as I understand) is set when there's a "parent" window. And Windows considers any "control" (like a button) a "window". So I don't know if that function we are using refers to those children controls and to identify them in the context of a parent normal window. The documentation is not really super clear to me...

**Comment by tiangolo** [comment]
The error I was trying to "solve" in the Travis build is this one: https://travis-ci.org/sindresorhus/active-win/jobs/226051518#L607

But now I see that [at the end](https://travis-ci.org/sindresorhus/active-win/jobs/226051518#L1144): `The command "npm install" exited with 0.`.

So I guess that maybe I'm trying to "solve" something that doesn't need to be solved as it is in the `optionalDependencies`. I'm sorry, I'm not an expert in Travis... yet.

It seems to me that the error is when `xo` is checking `windows.js`. And there's when it is being unable to find `fastcall` (although it is not needed for Linux / macOS).

How should I solve it?

Should I put all that from `windows.js` in a `try ... catch`? Or should I change `xo` configurations? Or... ?

**Comment by sindresorhus** [comment]
> So I guess that maybe I'm trying to "solve" something that doesn't need to be solved as it is in the optionalDependencies. I'm sorry, I'm not an expert in Travis... yet.

Yes. We can just ignore it. So remove the .travis.yml changes.

> Should I put all that from windows.js in a try ... catch? Or should I change xo configurations? Or... ?

Just put `/* eslint-disable import/no-unresolved */` at the top of `windows.js`. I've opened an issue about this: https://github.com/benmosher/eslint-plugin-import/issues/812

> Hmm, I know that the function we are using returns the thread that created the Window. Maybe we could create an ID based on the PID and thread ID. But if that thread creates more than one Window it wouldn't work as expectect, I think.

If a process opens two different windows, shouldn't those windows have two different numbers returned from `GetWindowLong(hwndChild, GWL_ID)`. Did you try opening multiple windows and see if it returns different numbers?

**Comment by tiangolo** [comment]
Great!

All changes done.

> If a process opens two different windows, shouldn't those windows have two different numbers returned from GetWindowLong(hwndChild, GWL_ID). Did you try opening multiple windows and see if it returns different numbers?

I tried having different windows created by different processes and all returned `0`. But I haven't tried a single process creating multiple Windows. I will try that. I think Firefox is still single process... I will try with it.

**Comment by tiangolo** [comment]
I just tested it. Firefox (at least the old version I had installed) is still single process. 

I opened 2 windows, both created by the same process (checked it with ProcessExplorer). But still, their "window ID" is `0`. For both of them.

**Comment by tiangolo** [comment]
I just tried concatenating the process ID and the thread ID that created the window. But it results that, for example that Firefox I have, both testing windows are created by the same process and same thread. So that didn't work either.

**Comment by sindresorhus** [comment]
I downloaded this to inspect the windows: http://www.nirsoft.net/utils/gui_prop_view.html

And it does look like windows have unique IDs:

screen shot 2017-04-27 at 21 08 47

screen shot 2017-04-27 at 21 10 42

Actually it seems both the handle and window IDs are unique numbers. No idea how we could get those though.

**Comment by tiangolo** [comment]
I have been reading the Windows docs about it. And even when enumerating all the windows, it seems that the functions only pass "handles" to those windows, and never an ID.

I won't paste here all the links to "possibly related" docs I found. But there's this long one, in where there's supposed to be a description of all the aspects on how "windows" work in Windows: https://msdn.microsoft.com/en-us/library/windows/desktop/ms632599(v=vs.85).aspx

The only mention I find of a window ID is in the context of child windows, that pertain to a parent window. You can search for "`A child window can have a unique integer identifier.`" to get to that section.

I'm out of ideas of where to search for a "window ID" in Windows. To me, they seem to think only in terms of passing handlers / pointers. :disappointed: 

---

I can think of hacks that coders could do to simulate a window ID, like hashing the PID and the window title. But still, if they had two windows created by the same process and with the same title it wouldn't work. As those tricks would be more "hackerish" I think we should let the library user (the coder) deal with them for their use case directly, knowing the risks and possible problems.

I don't see a clear and reliable way for us to provide that functionality here really. What do you think we should do?

**Comment by tiangolo** [comment]
:tada: Great idea. That seems to do it. 

The handle shown there seems to be the "memory address". It seemed strange to me, as I thought that it would create a new handle pointing to the same window for each process that asked for one, so I thought that each process would get a different handle and a different number. And in here: https://msdn.microsoft.com/en-us/library/windows/desktop/ms724469(v=vs.85).aspx it says that sometimes the same handle is used and sometimes it's shared.

But I just tested it locally, and it seems that for "windows" the handles are shared. I was able to get the same number in code that the one I was getting in that tool. :tada: 

Thanks for the great insight. Let me code it and I'll update.

**Comment by tiangolo** [comment]
Done! :white_check_mark: 

It is now returning the memory address as the window ID.

I tested it again, including the case of same process and thread but different windows (Firefox) and I still get a probably "unique" ID, that is the same as in the GUI tool you linked.

So, it seems to work! :tada:

**Comment by sindresorhus** [comment]
Yay. I'm glad we could resolve that.

**Comment by sindresorhus** [comment]
Thanks for working on this @tiangolo. This is a great addition :)

**Comment by tiangolo** [comment]
Thank you! :tada:

**Pull request merged**